DI container features

RefactorForge opened the pull request:
Imlementation of container logic with custom classes for 3 selected clustering algorithms
Provided also basic test cases

Kindly asking for review

Thank you

Commit:
feat: Adds implementation of custom classes for clustering algorithms
	- Implemented three custom classes based on according clustering algorithm,
	supports DI and provides function for clustering given data

Commit:
test: Adds tests for classes from clustering module
	- Implements basic testing for each custom class

Commit:
feat: Adds implementation of container class from Dependency Injector framework
	- Implemented custom container class from Dependency Injector framework
	- Container provides support for IO handling operations packed together with custom classes
	- Each custom algorithm class require explicit set of parameters (based on clustering algorithm)

Commit:
test: Adds tests for dependencies module
	- Implements basic test and mock of expeted behaviour of container created with inejcted parameters
	- Asserts different resutls for different configurations

tynaneu commented on app/clustering.py:
```diff
@@ -0,0 +1,270 @@
+from sklearn.cluster import DBSCAN, AgglomerativeClustering, KMeans
+
+
+class CKMeans:
```
there should be base class / interface for all cluster objects (see handlers)

RefactorForge replied:
Refactored

tynaneu commented on app/clustering.py:
```diff
@@ -0,0 +1,270 @@
+            compute_distances=compute_distances,
+        )
+
+    def cluster(self, data: ndarray) -> ndarray:
```
this method does not take sample_weights as other cluster methods, there more ways how to solve this:
1. all methods would take the argument and it can be None, when not None in this method, you can raise NotImplementedError, because this algorithm does not support it
2. do not use sample_weights in no custom implementation of clustering

RefactorForge replied:
Refactored.

tynaneu commented on tests/test_clustering.py:
```diff
@@ -0,0 +1,61 @@
+
+def test_ckmeans():
+    """Test CKMeans clustering."""
+    data = array([[1, 2], [1, 4], [1, 0], [10, 2], [10, 4], [10, 0]])
```
data are used in all tests - they should be in some fixture

RefactorForge replied:
Refactored.

tynaneu commented on tests/test_clustering.py:
```diff
@@ -0,0 +1,61 @@
+    )
+    labels = kmeans.cluster(data)
+    assert isinstance(labels, ndarray), "Output labels should be an ndarray."
+    assert len(labels) == len(data), "Each data point should have a label."
```
you do not test the functionality of the custom clustering - this test is an integration test. Please also write unit tests that check that the sci-kit learn method was called with given parameters, you mock for scikit_learn model - do not call it in unit tests

RefactorForge replied:
Refactored and separated from unit tests

tynaneu commented on app/dependencies.py:
```diff
@@ -0,0 +1,176 @@
+    """
+    Dependency Injection container for clustering and I/O operations.
+
+    This container centralizes configuration and
```
the docstring is longer than the actual code :) it is sufficient to use shorter docstrings that include only brief information about the class that helps the user to better understand the purpose of the class without seeing the actual implementation

RefactorForge replied:
Refactored and simplified.

tynaneu commented on tests/test_dependencies.py:
This test IMHO does not make any sense - if you want to test dependencies and creation of objects from config, you should create them without mocking and check that the created object is an instance of the right class and includes the right hyperparameters.
Testing return values of the algorithms does not make sense in dependency test - this should be in integration tests but without mocking.

RefactorForge replied:
Refactored.

tynaneu commented on tests/test_dependencies.py:
```diff
@@ -0,0 +1,252 @@
+from app.dependencies import ClusteringContainer
+
+
+class MockKMeans:
```
classes can be mocked using Mock() / MagickMock() or path from unittesting

RefactorForge replied:
Should be resolved.

Commit:
refactor: Changes structure of clustering module classes:
	- Implemented changes based of review comments for clustering module
	- Implemented abstract BaseCluster class for defining methods for subclasses and better docs organization
	- simlified docstrings

Commit:
refactor: Changes dependencies module:
	- simlified docstrings
	- removes io_handler variable from container to reflect changes of classes in clustering module

Commit:
test: Refactors and separates test_clustering test cases:
	- Implemented changes based of review comments for test_clustering
	- Separated integrations tests into file test_clustering_integration
	- Implemented test cases to  check that the sci-kit learn method was called with given parameters

Commit:
test: Refactors test_dependencies cases:
	- Implemented changes for test cases to actually test meaningful attributes (instance creations and if attributs are correct)

Acnologia7 commented:
@tynaneu 

Changes based on review comments should be implemented.
Kindly asking for new review.

Thank you
